sound/soc/sof/fw-file-layout.c

```diff
@@ -0,0 +1,355 @@
+						     GFP_KERNEL);
+		if (!out_layout->tplg_path)
+			goto err;
+		break;
+	default:
+		dev_err(dev, "Invalid firmware layout type: %d\n", layout_type);
+		return -EINVAL;
+	}
+
+	return 0;
+
+err:
+	sof_free_profile_strings(dev, out_layout);
+
+	return ret;
+}
+
+static int sof_test_fw_layout(struct device *dev,
+			      struct sof_fw_layout_profile *fw_layout)
+{
+	enum sof_ipc_type fw_ipc_type;
+	const struct firmware *fw;
+	const char *fw_filename;
+	const u32 *magic;
+	int ret;
+
+	fw_filename = kasprintf(GFP_KERNEL, "%s/%s", fw_layout->fw_path,
+				fw_layout->fw_name);
+	if (!fw_filename)
+		return -ENOMEM;
+
+	ret = firmware_request_nowarn(&fw, fw_filename, dev);
+	if (ret < 0)
+		return ret;
+
+	/* firmware file exists, check the magic number */
+	magic = (const u32 *)fw->data;
+	switch (*magic) {
+	case SOF_EXT_MAN_MAGIC_NUMBER:
+		fw_ipc_type = SOF_IPC_TYPE_3;
+		break;
+	case SOF_EXT_MAN4_MAGIC_NUMBER:
+		fw_ipc_type = SOF_IPC_TYPE_4;
+		break;
+	default:
+		dev_err(dev, "Invalid firmware magic: %#x\n", *magic);
+		ret = -EINVAL;
+		goto out;
+	}
+
+	if (fw_ipc_type != fw_layout->ipc_type) {
+		dev_err(dev,
+			"ipc type mismatch between file and profile: %d vs %d\n",
+			fw_ipc_type, fw_layout->ipc_type);
+		ret = -EINVAL;
+	}
+out:
+	release_firmware(fw);
+
+	return ret;
+}
+
+static int fw_layout_for_ipc_type(struct device *dev,
+				  const enum sof_ipc_type ipc_type,
+				  const struct sof_dev_desc *desc,
+				  const char *fw_path_postfix,
+				  struct sof_fw_layout_profile *out_layout)
+{
+	bool found = false;
+	int i, ret;
+
+	memset(out_layout, 0, sizeof(*out_layout));
+
+	for (i = 0; i < ARRAY_SIZE(layout_to_ipc_type_map); i++) {
+		if (layout_to_ipc_type_map[i] != ipc_type)
+			continue;
+
+		ret = setup_fw_layout_profile(dev, i, desc->vendor,
+					      desc->platform, fw_path_postfix,
+					      out_layout);
+		if (ret)
+			return ret;
+
+		ret = sof_test_fw_layout(dev, out_layout);
+		if (!ret) {
+			found = true;
+			break;
+		}
+
+		sof_free_profile_strings(dev, out_layout);
+	}
+
+	if (!found)
+		return -ENOENT;
+
+	dev_dbg(dev, "Default profile type for IPC type %d: %d\n", ipc_type, i);
+
+	return 0;
+}
+
+int sof_create_default_fw_layout_profile(struct device *dev,
+					 const enum sof_ipc_type ipc_type,
+					 const struct sof_dev_desc *desc,
+					 const char *fw_path_postfix,
+					 struct sof_fw_layout_profile *out_layout)
+{
+	int ret = -ENOENT;
+	int i;
+
+	memset(out_layout, 0, sizeof(*out_layout));
+
+	ret = fw_layout_for_ipc_type(dev, ipc_type, desc, fw_path_postfix,
+				     out_layout);
+	if (!ret)
+		return 0;
+
+	dev_warn(dev, "No default profile found for requested IPC type %d\n",
+		 ipc_type);
+
+	/*
+	 * No firmware file was found for the requested IPC type, check all
+	 * IPC types as fallback
+	 */
+	for (i = 0; i < SOF_IPC_TYPE_COUNT; i++) {
+		if (i == ipc_type || !(desc->ipc_supported_mask & BIT(i)))
+			continue;
+
+		ret = fw_layout_for_ipc_type(dev, i, desc, fw_path_postfix,
```

Comment:
So this is the magic bit that allows to choose IPC major version based on available installed FW.

Reply:
Yes, we could change the default IPC type to IPC4 for everything in pci-tgl.c and provide only selected platform's IPC4 firmware (where we have good coverage of topologies). That would work fine.
This would also allow the possibility (if ever needed) to change the location of the SOF firmware things to
/lib/firmware/sof/<vendor>/<ipc_type>/.../
if the profile is described in kernel.

But looks like this is an overshot, so I'll try to scale it back.